Add context to inventory plugins documentation (#72292)
ansible/ansible: docs/docsite/rst/plugins/inventory.rst

   [inventory]
   enable_plugins = host_list, script, auto, yaml, ini, toml

> If the plugin is in a collection, and isn't being picked up by the `auto` statement, you can append the fully qualified name:

Reviewer: ```suggestion
If the plugin is in a collection and is not being picked up by the `auto` statement, you can append the fully qualified name:
```

.. code-block:: ini

   [inventory]
   enable_plugins = host_list, script, auto, yaml, ini, toml, namespace.collection_name.inventory_plugin_name

> Or, if it's a local plugin, perhaps stored in `plugins/inventory/my_plugin.py`, you could reference it like this:

Reviewer 2: Maybe use the default path for local plugins instead https://docs.ansible.com/ansible/latest/dev_guide/developing_locally.html.

Author: Resolved in 8511d14

> Or, if it's a local plugin, perhaps stored in `~/.ansible/plugins/inventory/my_plugin.py`, you could reference it like this:

Reviewer: ```suggestion
Or, if it is a local plugin, perhaps stored in `~/.ansible/plugins/inventory/my_plugin.py`, you could reference it like this:
```

> Or, if it is a local plugin, perhaps stored in `~/.ansible/plugins/inventory/my_plugin.py`, you could reference it like this:

Reviewer 3: ```suggestion
Or, if it is a local plugin, perhaps stored in the path set by :ref:`DEFAULT_INVENTORY_PLUGIN_PATH`, you could reference it as follows:
```

   [inventory]
   enable_plugins = namespace.collection_name.inventory_plugin_name

If you want to be able to add a plugin from a collection to your other inventory methods, add it into the existing list:

.. code-block:: ini

   [inventory]

>    enable_plugins = host_list, script, auto, yaml, ini, toml, namespace.collection_name.inventory_plugin_name

Reviewer 4: i would note that if plugin author uses the `plugin: fqcn` convention this is not needed as the `auto` plugin should just load the appropriate plugin

Author: OK, I'll note that, instead of the convoulted route I took, you should instead make sure you're using the fqcn.

   [inventory]
   enable_plugins = host_list, script, auto, yaml, ini, toml, namespace.collection_name.inventory_plugin_name

> Or, if it is a local plugin, perhaps stored in `~/.ansible/plugins/inventory/my_plugin.py`, you could reference it like this:

Reviewer 4: i would not give this path but the 'setting' that cointains this and other paths  DEFAULT_INVENTORY_PLUGIN_PATH (link to it's doc page, which shows the defaults)

      - key: tags.Name
        prefix: tag_Name_
        separator: ""

>       # If you have a tag called "Role" which has the value "Webserver", this will add the group

Reviewer 2: This is demonstrating the same thing as the tags.Name example, so maybe we can improve the wording there.

Author: So, `tags.Name` provides an indicator of the role that the "separator" flag provides. Perhaps, instead, we should put something different in there - perhaps using some other tag like this:

```yml
- key: tags.Features
  prefix: Feature
  separator: "%"
```

And show the groups:

```
|--@Feature%Blue
|  |--abc123
|--@Feature%Green
|  |--def456
```

If this is an acceptable counter-offer, I'll make this change instead.

Reviewer 2: Oh, true. We shouldn't recommend using special characters in group names though. The updated examples look good to me, but it was pointed out to me that this page should try to be a simple overview, and these should probably go to the plugin-specific docs for constructed here https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/constructed.py#L26. You could add a comment to this page though, to see the constructed plugin for more examples if there isn't already one.

We might be able to clarify the wording for this comment - initially I read it as create groups for the tag "Role" if it has the value "Webserver", whereas the role_value groups are formed regardless of the value. But I'll leave that to the docs team.

Author: OK, so how about this, we yank all of the separator stuff out of here, so it just becomes:

```
- key: tags.Role
  prefix: role
```

(which, I think is generally a better and more accurate example than "this host was named something" as you don't tend to have lots of machines with the same name...)

Then, we fix up the constructed document with some of the examples from here (removing them from this as a result), perhaps including:

```
compose:
  ansible_host: public_ip_address is defined | ternary(public_ip_address, private_ip_address)
  some_inventory_wide_string: '"Yes, you need both types of quotes here"'
```

Reviewer: @s-hertel Could you please help here? Thanks in advance.

>       # Role_Webserver and add any hosts that have that tag assigned to it.

Reviewer 2: role_Webserver

Author: Fixed in 6a55691

      - key: tags.Role
        prefix: role
    groups:
      # add hosts to the group development if any of the dictionary's keys or values is the word 'devel'
      development: "'devel' in (tags|list)"
      # add hosts to the "private_only" group if the host doesn't have a public IP associated to it

>       private_only: "(network_interfaces[0].association | default({}) ).public_ip | default('') != ''"

Reviewer 2: As long as public_ip isn't sometimes an empty string, this could be
```suggestion
      private_only: "(network_interfaces[0].association | default({}) ).public_ip is defined"
```

Why not check the top level variable `public_ip_address` though?

Author: Ah, so let's cover off the "why not check the top level variable" - because I completely missed it! :) I looked for it on a node that didn't have a public IP, and it wasn't there, so I figured I'd have to "find it" somewhere else, and then went looking for where the attachment happened elsewhere - `network_interfaces`. Having now looked for it, I can clearly see it. D'oh.

Also yes, `is defined` versus `| default('') != ''` makes much more sense. I have hit issues with this in the dim and distant past, mostly with parsing environment variables, and so my "default reaction" is to go with the latter form rather than the former. Yes, I need to fix this one!

Author: Oh, an interesting quirk. `public_ip_address` is only defined where the first interface has a public ip defined. If you attach an EIP to the second or later interfaces, then it's not classed as that host's public ip address. This is totally outside the scope of this documentation change, but it gives me a bit more context for why I might have missed this before :)

    compose:
      # use a private address where a public one isn't assigned

>       ansible_host: ((network_interfaces[0].association | default({}) ).public_ip | default('') != '') | ternary((network_interfaces[0].association | default({}) ).public_ip, private_ip_address)

Reviewer 2: You don't need to use `|default({})` in the ternary filter since the public ip was found.

Author: Of course. Good spot.
